server/middleware.go

```diff
@@ -42,6 +43,49 @@ func JSONToHTTP(ep JSONEndpoint) http.Handler {
 	})
 }
 
+// ContextToHTTP is middleware func to convert ContextHandler to
+// an http.HandlerFunc.
+func ContextToHTTP(ctx context.Context, ep ContextHandler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Body != nil {
+			defer func() {
+				if err := r.Body.Close(); err != nil {
+					Log.Warn("unable to close request body: ", err)
+				}
+			}()
+		}
+
+		var (
+			cancel context.CancelFunc
+		)
+
+		ctx, cancel = context.WithCancel(ctx)
+
+		defer cancel()
+
+		ip, err := GetIP(r)
+
+		if err != nil {
+			LogWithFields(r).Warningf("unable to get IP: %s", err)
+		} else {
+			ctx = context.WithValue(ctx, "ip", ip)
+		}
+
+		if ip = GetForwardedIP(r); len(ip) > 0 {
+			ctx = context.WithValue(ctx, "forward-for-ip", ip)
+		}
+
+		id, err := GenerateRequestID()
```

Comment:
Other folks may start adding additional tracing via [go-kit/kit/tracing](https://github.com/go-kit/kit/tree/master/tracing#package-tracing) or some other tooling so we might not want to have this in here by default.

Would you mind abstracting this chunk with the RequestID into a middleware func? Then you can just add it into the ContextMiddleware method of your ContextService implementation.

Also, for the context keys, would you mind creating a separate type for them? I'd like us to match the pattern covered in the Go context blog post. (example: https://blog.golang.org/context/userip/userip.go)